### PR TITLE
DI-406 Edge cases for dentist ods

### PR DESCRIPTION
**Pull request by Corbisc**
## Link to JIRA Ticket

- https://nhsd-jira.digital.nhs.uk/browse/DI-406

## Description

This ticket adds a few new things.
There is a new generic log check step that will enable the user to specify a log group, a field, and a message to be verified. This will allow multiple tests to be moved across to this format, limiting the amount of test steps that are required.
There is also a new step that checks the opposite, that a certain log entry is not present, however this has required multiple requests to be made.
The new function in aws.py verifies that no logs are returned, and returns True, if nothing was found.
The new step runs the following:
An initial request to check logs for the correlation ID to find the function request ID.
A second request to ensure that the end step for the request ID has been found. This ensures that the Lambda has finished running.
A third request to verify that the log entry has not been found.

When running tests to verify performance change the 3 requests added ~20s to the runtime of the standard single request entry. This isn't too bad, as the negative check will be performed considerably less than the positive check. In the test environment results can be affected greatly by simultaneous test runs, which cannot be validated on my machine. Anyone reviewing should run these tests as part of a larger parallel test run.

An arg has also been added to the get_logs function so that the test can specify how long to retry for. This defaults to 32, which was the value set previously.

### Noteworthy Changes

- These are changes the reviewer should look out for

## Type of change

Test

## Testing

Please tick the testing that has been completed

- [ ] Unit tests
- [ ] Integration tests

## Developer Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have run the [code formatting checks](../README.md#code-quality)
- [ ] I have run the [code quality checks](../README.md#code-quality)
- [ ] New code meets [standards](https://nhsd-confluence.digital.nhs.uk/display/DI/DI+Ways+of+Working) agreed by the team
- [ ] Unit test code coverage is at or above 80%
- [ ] New and existing unit tests pass locally with my changes
- [ ] Tests have added that prove my fix is effective or that my feature works (Integration tests)
- [ ] I have made corresponding changes to the documentation
- [ ] I have cleaned down my environment (if created)

## Code Reviewer Checklist

- [ ] I have run the unit tests and they run correctly
- [ ] I can confirm the changes have been tested or approved by a tester
- [ ] I can confirm no remaining infrastructure is left over from this branch

### CODE REVIEW

File: test/integration/features/F002_Invalid_Change_Events.feature [hunk 1]
```diff
@@ -158,15 +158,26 @@ Feature: F002. Invalid change event Exception handling
     Then the Event Processor logs to splunk with report key "INVALID_OPEN_TIMES"
 
 @complete @cloudwatch_queries
-  Scenario: F002S021. Dentist Unmatched Service Type uses correct report key
+  Scenario Outline: F002S021. Dentist Unmatched Pharmacy and Service report keys
```
Comment (ed-sparkes) [inline test/integration/features/F002_Invalid_Change_Events.feature:161]: Where is the odscode now provided?

File: test/integration/features/F002_Invalid_Change_Events.feature [hunk 2]
```diff
@@ -158,15 +158,26 @@ Feature: F002. Invalid change event Exception handling
...
     Given a Dentist Changed Event is valid
-    And the Changed Event has ODS Code "FQG8101"
+    And the Changed Event has ODS Code "<ods_code>"
     When the Changed Event is sent for processing with "valid" api key
-    Then the Event Processor logs to splunk with report key "UNMATCHED_SERVICE_TYPE"
```
Comment (ed-sparkes) [inline test/integration/features/F002_Invalid_Change_Events.feature:165]: We are not actually checking its logged to splunk so maybe we should change the name of this to "... logs with report key ...."

File: test/integration/steps/utilities/aws.py [hunk 3]
```diff
@@ -61,11 +61,39 @@ def get_logs(query: str, event_lambda: str, start_time: Timestamp) -> str:
         counter += 1
         if response["results"] != []:
             logs_found = True
-        elif counter == 32:
+        elif counter == retrycount:
```
Comment (ed-sparkes) [inline test/integration/steps/utilities/aws.py:64]: Do some logs really take 32 * 20 seconds to arrive, 11 mins?

**Comment by ed-sparkes** [comment]
Looks good  Kit, 

Its still not ideal to be validated functionality purely on what the logs say or don't say but in the absence of anything better i think your plan for a negative log makes sense and will help us gain more coverage

**Review by mik3ola — APPROVED** [review]

**Pull request merged**